### DISCUSSION TITLE
Is it possible for Mojo 🔥 to communicate directly with the hardware without the operating system ❓💽💻🖥️

### DISCUSSION THREAD

**Discussion by Eprahim-taha** [modular/modular#1627] [Q&A]
C is a programming language designed by Dennis Ritchie at Bell Labs. C is very widely used, straightforward, and can be compiled to a number of platforms and operating systems. C is an imperative language, with a small number of keywords and a large number of mathematical operators. C is also a very low level programming language, which means it can communicate directly with hardware.

The question here is whether the Mojo programming language 🔥 has the ability to communicate with devices directly like the C language or not.  For example, programming a microcontroller to control a machine's motor.  Does Mojo have C abilities?  If the Mojo programming language does not have these capabilities that C possesses, then what is the point of it really!?

**Comment by Sharktheone** [comment]
Yes, there is a way to do FFI currently. You can use the [`external_call`](https://docs.modular.com/mojo/stdlib/sys/intrinsics.html#external_call) function from the std lib to do that.

However, there isn't any good documentation on how to link against static libs or shared libraries, but [here](https://github.com/ihnorton/mojo-ffi/tree/ffi-demo) is a repository that provides some examples.

**Reply by Eprahim-taha** [reply]
Thank 👏

**Reply by Eprahim-taha** [reply]
What I mean is, do I have Mojo 🔥 the ability to communicate with devices without an operating system, for example direct communication with the processor and memory without an operating system? this is what I meant. Thank you for your interest..

**Reply by Sharktheone** [reply]
Please read the answer on your other discussion on this #1561 and #1461. It is still the same answer and it probably stays like this for a while.

**Reply by Eprahim-taha** [reply]
Are you an employee at modular 🤔

**Reply by Sharktheone** [reply]
No, just answering GH questions

**Reply by Eprahim-taha** [reply]
Thanks for your time
